BDMS-221-225: Polymorphic DataProvenance model

ksmuczynski opened the pull request:
<!--
- BDMS-221-225: Polymorphic DataProvenance model
- Keep sections short and scannable.
-->
### **Why**
_This PR addresses the following problem / context:_
- Centralizes provenance metadata for foundational/static records to improve data traceability.
- Enables models to associate provenance information efficiently using reusable mixins.
- Refines location and thing models to support provenance tracking and standardized metadata.
- The purpose of this PR is to make these intermediate updates available to @jacob-a-brown as we continue to co-work on the main BDMS-221 task.

### **How**
_Implementation summary - the following was changed / added / removed:_
- Added new DataProvenance model in db/data_provenance.py to centralize provenance metadata for foundational/static records.
- Introduced DataProvenanceMixin in db/base.py for reusable polymorphic relationships, enabling models to associate provenance efficiently.
- Added DataProvenanceMixin to the Location and Thing models to support provenance tracking for location and thing records.
- Enhanced field definitions, relationships, and documentation in all affected files for clarity and maintainability.

### **Notes**
_Any special considerations, workarounds, or follow-up work to note?_
- Lexicon categories and terms for provenance fields need to be added to cover all relevant controlled vocabularies.
- Further integration with other models may be required as provenance tracking is extended.

ksmuczynski committed:
feat: add DataProvenance model and enhance base mixins

The current schema lacks a way to store and track provenance (origin) data across the database.

Created db/data_provenance.py with a polymorphic DataProvenance model for tracking foundational metadata across tables.
Added mixin DataProvenanceMixin to db/base.py for reusable polymorphic relationships.
Improved documentation and comments in db/base.py for mixins and helper functions.

ksmuczynski committed:
feat: add DataProvenanceMixin for polymorphic provenance tracking

Introduced DataProvenanceMixin to the `Thing` and `Location` models to enable reusable, efficient, polymorphic relationships to the DataProvenance table.

jacob-a-brown commented on db/data_provenance.py:
```diff
@@ -0,0 +1,103 @@
+    # --- Polymorphic Parent Relationships (Internal) ---
+    # These are view-only relationships used by the 'target' property below.
+    # They tell SQLAlchemy exactly how to find the specific parent record for a given child.
+    _thing_target: Mapped["Thing"] = relationship(
+        "Thing",
+        primaryjoin="and_(foreign(DataProvenance.target_id) == Thing.id, DataProvenance.target_table == 'Thing')",
+        viewonly=True,
+    )
+    _location_target: Mapped["Location"] = relationship(
+        "Location",
+        primaryjoin="and_(foreign(DataProvenance.target_id) == Location.id, DataProvenance.target_table == 'Location')",
+        viewonly=True,
+    )
```
because the table uses `snake_case` in the database I think that it should be `DataProvenance.target_table == 'thing')` (and the same with `location`)

jacob-a-brown commented on db/base.py:
```diff
@@ -210,6 +210,25 @@ def permissions(self):
+            viewonly=True,
+        )
+
+
```
To future proof this and allow for `CamelCase` to `snake_case` models --> table names this should be moved to `db/data_provenance.py` and be

```python
from db.base import pascal_to_snake
from sqlalchemy import and_
from sqlalchemy.orm import declared_attr, relationship, foreign


class DataProvenanceMixin:
    """
    Mixin for models that can have data provenance records (e.g., Thing, Location).
    It automatically creates a polymorphic One-to-Many relationship to the
    DataProvenance table.
    """

    @declared_attr
    def data_provenance(cls):
        # One-to-Many polymorphic relationship
        return relationship(
            "DataProvenance",
            primaryjoin=and_(
                cls.id == foreign(DataProvenance.target_id),
                DataProvenance.target_table == pascal_to_snake(cls.__name__)
            ),
            lazy="selectin",
            viewonly=True,
        )
```

jacob-a-brown commented on db/data_provenance.py:
```diff
@@ -0,0 +1,103 @@
+    # TODO: Values from the following NMAquifer tables should be included as terms in the lexicon:
+    #  'LU_AltitudeMethod','LU_CoordinateMethod'.
+    collection_method: Mapped[str] = lexicon_term(
+        nullable=True,
+        comment="Indicates the method used to collect the data (e.g., 'GPS - Survey Grade').",
+    )
+    # TODO: Values from the following NMAquifer tables should be included as terms in the lexicon: 'LU_CoordinateAccuracy'.
+    accuracy_value: Mapped[float] = mapped_column(
+        nullable=True, comment="A numeric value representing the data's accuracy."
+    )
+    # TODO: Values from the following NMAquifer tables should be included as terms in the lexicon: 'LU_CoordinateAccuracy'.
+    accuracy_unit: Mapped[str] = lexicon_term(
+        nullable=True,
+        comment="The unit for the `accuracy_value` (e.g., 'meters', 'feet').",
+    )
```
these fields are used for a small subset of fields in a subset of tables. should these move to those tables, rather than be here? since they won't apply to a number of fields for which `DataProvenance` will be used.

ksmuczynski replied:
Maybe, but I saw this as one of the advantages of the Provenance model - all the sparse, optional, and evolving metadata is organized in one central place. If we move these fields to the Location table we'd have to add even more fields (`coordinate_accuracy`, `coordinate_collection_method`, `coordinate_accuracy_value`, `coordiante_accuracy_unit`, plus the same ones for elevation). I think storing this type of metadata is more efficient with the `DataProvenance` table, but will let @jirhiker weigh in, too.

jirhiker replied:
It seems appropriate to store these fields here.  We can reevaluate later if user requirements dictate

ksmuczynski committed:
refactor: refine polymorphic parent relationships.

The database tables are snake_case, so for consistency and ease of debugging, the `target_table` values should also use snake_case.

Refined the _thing_target and _location_target relationships to ensure DataProvenance.target_table uses snake_case ('thing', 'location') for the target table name.

ksmuczynski committed:
refactor: move DataProvenanceMixin to data_provenance.py and refactor for class-level usage

- Relocated DataProvenanceMixin from base.py to data_provenance.py for better modularity and provenance management.
- Refactored mixin to use cls in @declared_attr for proper class-level relationship definition.

ksmuczynski committed:
refactor: Update lexicon and `enums.py` with DataProvenance related information.

- Added new `origin_source` and `collection_method` categories and terms.
- Added 'meters' as a term associated with the `unit` category.
- Added `OriginStatus` to `enums.py`.

ksmuczynski committed:
refactor: Update lexicon and `enums.py` with DataProvenance related information.

- Added new `origin_source` and `collection_method` categories and terms.
- Added 'meters' as a term associated with the `unit` category.
- Added `OriginStatus` to `enums.py`.

jacob-a-brown committed:
Merge branch 'bdms-221' into kas-bdms-221-225-core-well-info-models-schemas
